std/array.d

```diff
@@ -778,39 +778,64 @@ private auto arrayAllocImpl(bool minimallyInitialized, T, I...)(I sizes) nothrow
 }
 
 // overlap
-/*
-NOTE: Undocumented for now, overlap does not yet work with ctfe.
+
+
+/++
 Returns the overlapping portion, if any, of two arrays. Unlike $(D
-equal), $(D overlap) only compares the pointers in the ranges, not the
-values referred by them. If $(D r1) and $(D r2) have an overlapping
-slice, returns that slice. Otherwise, returns the null slice.
-*/
-auto overlap(T, U)(T[] r1, U[] r2) @trusted pure nothrow
-if (is(typeof(r1.ptr < r2.ptr) == bool))
+equal), $(D overlap) only compares the pointers and lengths in the
+ranges, not the values referred by them. If $(D r1) and $(D r2) have an
+overlapping slice, returns that slice. Otherwise, returns the null
+slice.
++/
+
+auto overlap(T, U)(T[] a, U[] b) @trusted
+if (is(typeof(a.ptr < b.ptr) == bool))
 {
-    import std.algorithm.comparison : min, max;
-    auto b = max(r1.ptr, r2.ptr);
-    auto e = min(r1.ptr + r1.length, r2.ptr + r2.length);
-    return b < e ? b[0 .. e - b] : null;
+    import std.algorithm.comparison : min;
+
+    alias RetPtr = CommonType!(T*, U*);
+
+    if (a.ptr <= b.ptr && b.ptr < a.ptr + a.length)
```

Comment:
```
@trusted auto makeDanger()
{   return (new int)[0 .. 750];
}

@safe void main()
{   import std.stdio;

    auto dangerousArray = makeDanger;
    dangerousArray = dangerousArray[100 .. $];
    dangerousArray[0].writeln;  //R.I.P D-man
}
```
Even built-in functions trust the array bounds, the above is not caught. Memcpy is a bit different because you can give a wrong length for it without the compiler noticing it. But there is not, or should not be, a way to create invalid arrays in safe code.

But this debate is getting lenghty. If this did not convince you, I'd say we should let Andrei decide.